engine-transactions/src/eip_7702.rs

```diff
@@ -0,0 +1,262 @@
+        for tuple in &self.authorization_list {
+            s.begin_list(6);
+            s.append(&tuple.chain_id);
+            s.append(&tuple.address);
+            s.append(&tuple.nonce);
+            s.append(&tuple.parity);
+            s.append(&tuple.r);
+            s.append(&tuple.s);
+        }
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub struct SignedTransaction7702 {
+    pub transaction: Transaction7702,
+    /// The parity (0 for even, 1 for odd) of the y-value of a secp256k1 signature.
+    pub parity: u8,
+    pub r: U256,
+    pub s: U256,
+}
+
+impl SignedTransaction7702 {
+    const TRANSACTION_FIELDS: usize = 13;
+
+    pub fn sender(&self) -> Result<Address, Error> {
+        let mut rlp_stream = RlpStream::new();
+        rlp_stream.append(&TYPE_BYTE);
+        self.transaction.rlp_append_unsigned(&mut rlp_stream);
+        let message_hash = aurora_engine_sdk::keccak(rlp_stream.as_raw());
+        ecrecover(
+            message_hash,
+            &super::vrs_to_arr(self.parity, self.r, self.s),
+        )
+        .map_err(|_e| Error::EcRecover)
+    }
+
+    pub fn authorization_list(&self) -> Result<Vec<Authorization>, Error> {
+        if self.transaction.authorization_list.is_empty() {
+            return Err(Error::EmptyAuthorizationList);
+        }
+        let mut authorization_list = Vec::with_capacity(self.transaction.authorization_list.len());
+        // According to EIP-7702 we should validate each authorization. We shouldn't skip any of them.
+        // And just put `is_valid` flag to `false` if any of them is invalid. It's related to
+        // gas calculation, as each `authorization_list` must be charged, even if it's invalid.
+        // The exception is invalid signature, as it's indicate wrong transaction.
+        for auth in &self.transaction.authorization_list {
+            // Validate the signature, as in tests it is possible to have invalid signatures values.
+            let v = auth.parity.0;
+            if !(v[0] < u64::from(u8::MAX) && v[1..4].iter().all(|&elem| elem == 0)) {
+                return Err(Error::InvalidAuthorizationSignature);
+            }
+            // Value `v` shouldn't be greater then 1
+            let v = u8::try_from(v[0]).map_err(|_| Error::InvalidAuthorizationSignature)?;
+            if v > 1 {
+                return Err(Error::InvalidAuthorizationSignature);
+            }
+            // EIP-2 validation
+            if auth.s > SECP256K1N_HALF {
+                return Err(Error::InvalidAuthorizationSignature);
+            }
+
+            // According to EIP-7702 step 1. validation, we should verify is `chain_id` is related to current network.
+            // We just ensure that `chain_id` is not zero, as it's possible different `chain_id` and
+            // `chain_id == 0` is related to Ethereum `mainnet`, that's not correct for Aurora.
+            let mut is_valid = !auth.chain_id.is_zero();
```

Comment:
This check does not look right to me. The documentation says `Verify the chain id is either 0 or the chain's current ID`. So 0 is always allowed regardless of the current chain and any non-zero value must match the current chain id. This check does neither of these things because it always rejects 0 and accepts any non-zero value.

Instead I think the expected chain id should be passed as an argument to this function and the check should be `auth.chain_id.is_zero() || auth.chain_id == expected_chain_id`.

Reply:
I know about `chain_id = 0`. We intended to simplify it as we do not pass `chain_id` to `NormalizedTransaction`.
it. I'm agree it doesn't look like a good idea and it should be strictly related to EIP.

Reply:
Do you have an idea how to manage `current_chain_id` for:
```
impl TryFrom<EthTransactionKind> for NormalizedEthTransaction {
    type Error = Error;

    fn try_from(kind: EthTransactionKind) -> Result<Self, Self::Error> { ... }
```

Reply:
Hmm I see that doesn't quite work. Perhaps the validation logic should be separate from constructing `NormalizedEthTransaction`. Maybe it should exist as part of the `submit` function logic after the transaction is parsed.

Reply:
Agree.

Reply:
Solved: 976248fed50c0526337ea8e945021a72a79ca66e